proposals/0243-character-literals.md

```diff
@@ -0,0 +1,107 @@
+
+ ## Motivation
+
+ A pain point of using characters in Swift is they lack a first-class literal syntax. Users have to manually coerce string literals to a `Character` or `Unicode.Scalar` type using `as Character` or `as Unicode.Scalar`, respectively. Having the collection share the same syntax as its element also harms code clarity and makes it difficult to tell if a double-quoted literal is being used as a string or a character in some cases.
+
+ While the motivation for distinguishing between `String` and `Character` literals mostly consists of ergonomic and readability concerns, doing so would also bring Swift in line with other popular languages which do make this syntactic distinction, and facilitates a subsequent effort to improve support for low-level `UInt8`/`Int8` buffer processing tasks common in parsers and codecs.
+
+ ## Proposed solution
+
+ We propose to adopt the `'x'` as an alternative syntax for all textual literal types up to and including `ExtendedGraphemeClusterLiteral`, but not including `StringLiteral`. These literals will be used to express `Character`, `Unicode.Scalar`, and types like `UTF16.CodeUnit` in the standard library. These literals would have a default type of `Character`, as `Character` is the preferred element type of `String`. In addition where the character literal is a single ASCII code point conversions to an integer value are made available using a new `ExpressibleByASCIILiteral` conformance.
+
+ Use of single quotes for character/scalar literals is highly precedented in other languages, including C, Objective-C, C++, Java, Elm, and Rust, although different languages have slightly differing ideas about what a “character” is. We choose to use the single quote syntax specifically because it reinforces the notion that strings and character values are different: the former is a sequence, the later is an element (though a single element can itself be a `String`). Character types also don’t support string literal interpolation and can be obtimized, which is another reason to move away from double quotes.
+ 
+ Advantages for a developer to migrate to the single quote distinction:
+ 
+  * Differentiate in the source when a literal is intended to be a Character or UnicodeScalar
+  * Distinct default type of `Character` making available that type's methods and properties.
+  * Compile time best-effort check that the literal is in fact a single Character/Unicode grapheme.
+
+ Improvements to the new implementation over that previously reviewed:
+ 
+  * SingleQuoted literals have their own new `ExpressibleBy` marker protocols preventing source breaking changes to the use of double quoted literals in existing source.
+  * Distinct protocol for ASCII literals further localising the more contentious integer conversions.
```

Comment:
newspeak. reword and expand this a bit to explain exactly how this protocol benefits us and which integer conversions it supports vs disallows